### ISSUE TITLE
emit signal on partial oom

### ISSUE THREAD

**Issue by hsitter** [systemd/systemd#35067]
### Component

systemd

### Is your feature request related to a problem? Please describe

At KDE we'd like to know when a process gets oomd, and which cgroup/unit it belonged to, even when that cgroup/unit is still active.

### Describe the solution you'd like

DBus signal or something similar (ideally on the session bus).

### Describe alternatives you've considered

We currently already listen to property changes on units (Result==oom-kill) but those do not come into play when only part of a unit got oomd. For instance we have our terminal app split into different scopes and only some commands in those may get oomd (e.g. think of stress-ng running in a terminal tab -- strees-ng would get oomd but not the tab)

```
│   │ ├─app.slice
│   │ │ ├─app-org.kde.konsole-41769.scope
│   │ │ │ ├─main.scope
│   │ │ │ │ └─41769 /usr/bin/konsole
│   │ │ │ ├─tab(42253).scope
│   │ │ │ │ ├─42253 /bin/bash
│   │ │ │ │ └─42735 htop
│   │ │ │ └─tab(41807).scope
│   │ │ │   └─41807 /bin/bash
```

### The systemd version you checked that didn't have the feature you are asking for

256

**Comment by Sodivad** [comment]
> huh, what is konsole doing there? it uses .scope as suffix for its own subcgroups? uh, that's confusing... 

FWIW this is what gnome-terminal is doing

```
│   │ │ ├─app-org.gnome.Terminal.slice
│   │ │ │ ├─vte-spawn-393df4de-d554-412a-9e2c-d6c9bdaf576e.scope
│   │ │ │ │ └─44030 zsh
│   │ │ │ ├─vte-spawn-3b9c1eb4-97b9-407c-a379-7ec369563703.scope
│   │ │ │ │ └─43939 zsh
│   │ │ │ ├─vte-spawn-94e6f26a-a2f1-460d-b7d5-955116dd1080.scope
│   │ │ │ │ └─44897 zsh
│   │ │ │ ├─vte-spawn-37d8e598-2d9e-481b-9eb9-4c3886e47eb7.scope
│   │ │ │ │ ├─44969 zsh
│   │ │ │ │ └─45005 htop
│   │ │ │ ├─vte-spawn-43688849-9fa1-45ae-8a5b-20589d71eacc.scope
│   │ │ │ │ └─44109 zsh
│   │ │ │ ├─vte-spawn-f15576bb-6e2f-473a-9a47-f9834e9e6dd3.scope
│   │ │ │ │ └─44934 zsh
│   │ │ │ ├─gnome-terminal-server.service
│   │ │ │ │ └─43932 /usr/libexec/gnome-terminal-server
│   │ │ │ ├─vte-spawn-481500ca-b39a-4ada-8adc-32b151f6ba06.scope
│   │ │ │ │ └─43988 zsh
│   │ │ │ └─vte-spawn-816b179d-1e66-4ef7-a61a-c37700f87764.scope
│   │ │ │   └─44849 zsh
```

> oomd's "Killed" dbus signal, isn't that what you are looking for?

Potentially yes but not every distor/user is using oomd so it might just be  the kernel killing processes

**Comment by poettering** [comment]
> FWIW this is what gnome-terminal is doing

yeah, that's quite different: they allocate a slice with systemd and scopes within them also via systemd. as i read the konsole tree they ask for delegation and then manage everything themselves.

I think gnome-terminal/vte are doing it right here, they'll get per-scope oom events and everything. if konsole would do the same, this should fix their issue too.

**Comment by Sodivad** [comment]
> I think gnome-terminal/vte are doing it right here, they'll get per-scope oom events and everything. if konsole would do the same, this should fix their issue too.

I tested gnome-terminal and indeed when a process is first oomedI see systemd sending signals that the Unit is now failed and goot oom-killed. However this is not what's happened - the unit continues to exist and other processes inside it like the shell live. If you start another process in this scope that gets oomed there's no further signaling from systemd (since the unit is already failed?)
